add sagemaker studio data engineering sessions and extensions package recipes

Jas0nch opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.


These packages are build as `noarch:python` and will build fine in the feedstock because they are linux-only packages. They are not supposed to run/build in Windows

Jas0nch commented:
I, @Jas0nch , confirm I am willing to maintain this package

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/sagemaker-studio-dataengineering-extensions/meta.yaml```, ```recipes/sagemaker-studio-dataengineering-sessions/meta.yaml```) and found it was in an excellent condition.

freddylyu173 commented:
I, @freddylyu173 , confirm I am willing to maintain this package.

Jas0nch commented:
@conda-forge/help-python - the PR is ready for review. Thank you!

@conda-forge-admin, please ping conda-forge/help-python

Jas0nch commented:
These packages was built in noarch:python and does not support Windows

Jas0nch commented:
@conda-forge/help-python, ready for review!

@conda-forge-admin, please ping conda-forge/help-python

Jas0nch commented:
@ocefpaf would you mind help to review this PR? Also, what should be the steps if I want to be a reviewer of staged-recipes?

Jas0nch commented:
These packages are build as noarch:python and will build fine in the feedstock because they are linux-only packages. They are not supposed to run/build in Windows environment.

The packages requires a bash script to be executed during building process, that one works in linux/macos but failed in Windows as we don't provide a Windows version bat script for it. That's why it's failing in Windows and that's expected.

dhirschfeld commented on recipes/sagemaker-studio-dataengineering-extensions/post-link.sh:
```diff
@@ -0,0 +1,11 @@
+cd $PREFIX/lib/python$PYTHON_VERSION/site-packages/$PYTHON_PKG_NAME/sagemaker_data_explorer && pip install .
+cd $PREFIX/lib/python$PYTHON_VERSION/site-packages/$PYTHON_PKG_NAME/sagemaker_connection_magics_jlextension && pip install .
+cd $PREFIX/lib/python$PYTHON_VERSION/site-packages/$PYTHON_PKG_NAME/sagemaker_ui_doc_manager_jl_plugin && pip install .
+cd $PREFIX/lib/python$PYTHON_VERSION/site-packages/$PYTHON_PKG_NAME/sagemaker_studio_theme && pip install .
```
Someone from core might have to decide if this is acceptable.

Mixing `pip` and `conda` packages is often a recipe for corrupting ones environment.

Jas0nch replied:
can you help to ping someone from core team?

This is some sub packages with different functionality and we want it to ship in one conda package.

dhirschfeld replied:
The usual way would be to package each as `conda` packages and add them as dependencies.

Jas0nch replied:
we don't want package each, in the long run we will have them merged as one package. This post-link is just a temp workaround and will be removed in the future.

@conda-forge/core, can someone from core team help to review this?

dhirschfeld replied:
The problem is that `pip` installing into a `conda` environment can break that environment if `pip` decides it wants to clobber any `conda` installed files.

If you really do want to install 7 packages as one, you'd probably be better off downloading the source for each [in their own folder](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#destination-path) and then run ` {{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv` in each of the source folders.

hmaarrfk replied:
yeah this is not ok to do in a post-link step. all this kind of installation should happen during install.

hmaarrfk replied:
All these extensions seem like they should have already been installed.

mariusvniekerk replied:
If you want to just install 7 packages why not just do this at build time instead of at link time?

Jas0nch replied:
Thanks for your responses. 

These suggestions are super helpful!! I realized they are already installed and I don't need a post-link script to install them again. I'll have a revision to remove this script soon.

Jas0nch commented:
@conda-forge/core need help from core team.

dhirschfeld commented:
You can ping core by mentioning @conda-forge/core

Jas0nch commented:
Thanks for @dhirschfeld 's help on this PR, @conda-forge/core this PR may need someone from core team to review. Can someone take a look at this? I know you guys must be super busy, any attention is appreciated.

To summarize, this PR contains a recipe/package consist of multiple sub packages, those sub packages are developed independently and we are working on merging them together. Meanwhile, we want this package can be published to conda-forge. Therefore, we have a post-link script to pip install these sub packages as a temp workaround. So we are gathering some input from @conda-forge/core to decide if this is acceptable.

Some existing post-link script using pip install:

https://github.com/zsinnema/minian-feedstock/blob/64c7992a85e991e8f86810dc438d46d8671957e7/recipe/post-link.sh#L2

https://github.com/conda-forge/jupyterlab-lego-boost-feedstock/blob/main/recipe/post-link.sh#L1

hmaarrfk commented on recipes/sagemaker-studio-dataengineering-extensions/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+
+build:
+  noarch: python
+  script: ./src/sagemaker_studio_dataengineering_extensions/scripts/pypi_pre_build.sh && {{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv 
```
i'm not sure what this script does, but it feels suspect can you link us to the source of this script so I can see what it does?

Jas0nch replied:
Sure. 

Basically this script is to change pyproject.toml file to remove some AWS internal hatch plugin, which is required in AWS customized hatch build system but not for public hatch. I'm planing to remove this in the next revision as well.

You could download the source from https://pypi.org/project/sagemaker-studio-dataengineering-extensions/#sagemaker_studio_dataengineering_extensions-1.0.5.tar.gz.

hmaarrfk reviewed (changes requested):
you should not be using pip during a post-link and pre-unlink step.

Those are really for edge cases where one MUST define environment variables, but those are pretty rare for pure python pythons

Jas0nch commented:
@hmaarrfk I have updated a version without the post-link scripts, but the Azure pipeline fails, when I run `pixi run build-linux` in my local laptop, it can be successfully built, do you know what could be the reason?

The error msg from Azure pipeline is 

```
  ➤ YN0000: ┌ Resolution step
  ➤ YN0061: │ abab@npm:2.0.6 is deprecated: Use your platform's native atob() and btoa() methods instead
  ➤ YN0061: │ glob@npm:7.1.7 is deprecated: Glob versions prior to v9 are no longer supported
  ➤ YN0061: │ eslint@npm:8.57.1 is deprecated: This version is no longer supported. Please see https://eslint.org/version-support for other options.
  ➤ YN0061: │ inflight@npm:1.0.6 is deprecated: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
  ➤ YN0061: │ @humanwhocodes/config-array@npm:0.13.0 is deprecated: Use @eslint/config-array instead
  ➤ YN0061: │ rimraf@npm:3.0.2 is deprecated: Rimraf versions prior to v4 are no longer supported
  ➤ YN0061: │ @humanwhocodes/object-schema@npm:2.0.3 is deprecated: Use @eslint/object-schema instead
  ➤ YN0002: │ @jupyterlab/services@npm:7.3.5 doesn't provide react (p67edc), requested by @jupyterlab/settingregistry
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide @egjs/hammerjs (p40090), requested by vis-timeline
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide @egjs/hammerjs (pf6164), requested by vis-util
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide component-emitter (p458ab), requested by vis-timeline
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide component-emitter (p0e0f9), requested by vis-util
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide plotly.js (pa8995), requested by react-plotly.js
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide propagating-hammerjs (p026c5), requested by vis-timeline
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide uuid (p1eca3), requested by vis-data
  ➤ YN0002: │ sagemaker_sparkmonitor@workspace:. doesn't provide uuid (p95bd3), requested by vis-timeline
  ➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
  ➤ YN0000: └ Completed in 10s 325ms
  ➤ YN0000: ┌ Post-resolution validation
  ➤ YN0028: │ The lockfile would have been modified by this install, which is explicitly forbidden.
  ➤ YN0000: └ Completed
  ➤ YN0000: Failed with errors in 10s 402ms

   stderr is
  error: subprocess-exited-with-error
```

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/sagemaker-studio-dataengineering-extensions/meta.yaml```, ```recipes/sagemaker-studio-dataengineering-sessions/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/sagemaker-studio-dataengineering-extensions/meta.yaml**:

 * ❌ `noarch` packages can't have skips with selectors. If the selectors are necessary, please remove `noarch: python`.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13670444303](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13670444303). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/sagemaker-studio-dataengineering-sessions/meta.yaml```, ```recipes/sagemaker-studio-dataengineering-extensions/meta.yaml```) and found it was in an excellent condition.

Jas0nch commented:
> @hmaarrfk I have updated a version without the post-link scripts, but the Azure pipeline fails, when I run `pixi run build-linux` in my local laptop, it can be successfully built, do you know what could be the reason?
> 
> My guess is because the CI=azure was set when running in Azure pipeline, and that triggers yarn immutable install, do you know if we could find a way to prevent that?
> 

The reason is because of the CI flag was set, I'm able to fix this issue by having `jlpm install --no-immutable` when building the artifact.

I have addressed all of the comments, can I have someone to review this again?

mariusvniekerk commented:
Looks like this has windows build issues since the build script is directly invoking `rm` which is no present on windows

````
src/sagemaker_studio_dataengineering_extensions/scripts/build_hooks.py.BuildException: command cd src/sagemaker_studio_dataengineering_extensions/sagemaker_spark_monitor_widget && rm -rf .npmrc .yarnrc.yml && npm install failed with return_code 1
   stdout is
   stderr is 'rm' is not recognized as an internal or external command,
  operable program or batch file.
```

Should windows be excluded?

hmaarrfk commented:
If python is calling subprocess rm, this seems like it should be patched upstream….

Jas0nch commented:
> If python is calling subprocess rm, this seems like it should be patched upstream….

I don’t quite get that. What does it mean by patched upstream?

hmaarrfk commented:
if this is the future of `pyproject.toml` i don't want to be a part of it.... this build script looks like a nightmare. Basically python calling "shell"....
[build_hooks.py.txt](https://github.com/user-attachments/files/19099868/build_hooks.py.txt)

Technically we don't have a requirement that noarch packages pass on windows. So @mariusvniekerk has the power to "merge" now and get your package through.

I'm going to stay out of this one. Please ping me directly if you need anything, but I think you and Marius can take it from here.

> What does it mean by patched upstream?

typically projects have a github/gitlab page where you can submit patches to help improve their build process. There is no reason why they can't use the equivalent python command for `rm`.....

Jas0nch commented:
@hmaarrfk, thanks for the feedbacks. We have some restrictions here and it may make more sense to you if you see the full picture.

> Basically python calling "shell".... [build_hooks.py.txt](https://github.com/user-attachments/files/19099868/build_hooks.py.txt)

This package is a JupyterLab Extension, it involves some nodejs scripts in the build process. That's why we have so many shell commands to run in the build_hook.py. Please let me know if there's a way to trigger nodejs build from python directly, we would love to improve our build process.


> I'm going to stay out of this one. Please ping me directly if you need anything, but I think you and Marius can take it from here.

I saw you have requested changes, I'm not sure if we could proceed without your approval.... I'm not familiar with the settings of this repository so I could be wrong.

ref: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request

> The Request changes option is purely informational and will not prevent merging unless a ruleset or classic branch protection rule is configured with the "require a pull request" option. If configured and a collaborator with admin, owner, or write access to the repository submits a review requesting changes, the pull request cannot be merged until the same collaborator submits another review approving the changes in the pull request.

hmaarrfk commented:
> no longer does terribly scary things in link scripts

This was the main thing i wanted.

hmaarrfk commented:
> This package is a JupyterLab Extension, it involves some nodejs scripts in the build process. [...] Please let me know if there's a way to trigger nodejs build from python

I'm honestly not sure. Part of my frustration with `pypa` is that they have pushed people away from setup.py without listenning to our concerns. This has pushed you to create these workaround that I don't truely understand.

If you were targetting only conda, I would say you can:

1. Run the `npm` commands directly in bash/bat
2. Run the `pip install` without invoking any external libraries.

But instead, you are confined to working within the constraints setup by PyPi to reach your large userbase, which confines you to workarounds such as the one you have found.

Adding conda compatibility increases the size of your test matrix.

I don't envy your situation and am sympathetic to it.

Good luck!

Jas0nch commented:
Thanks for the help from Marius and Mark.

Yeah I don't like this kind of workaround either.....
